### ISSUE TITLE
base.List.map (etc.) is O(n log n) time instead of O(n) because it uses `##List.at` n times

### ISSUE THREAD

**Issue by aryairani** [unisonweb/unison#2107]
What's the plan for this sort of thing: new builtins for `map`, `flatMap`, `foldLeft`, `foldRight`? Something clever in the runtime that detects access patterns?

**Comment by aryairani** [comment]
Maybe we add builtins for `List.uncons` and `List.unsnoc` and then use them instead of `List.at` to implement the others in Unison.

**Comment by pchiusano** [comment]
We don't need any new builtins I don't think, and builtins that are higher-order functions are hard to do if you want them to be ability-polymorphic (which you definitely want `List.map` and these other functions to be). So I wouldn't go that route.

Important correction: in a finger tree, access to index `i` takes `log j` time where `j` is distance from the start or end of the list (whichever is closer). So the runtime is more like `log ((n/2)!)` which is better than `n log n`. That's still not linear - [plot](https://www.wolframalpha.com/input/?i=%28sum+%28log+n%29%2C+n+%3D+1+to+n%2F2%29) but it's not too shabby.

You could implement `List.map` and friends using pattern matching, walking over the list left to right. That would be linear time but involves more allocation since it's rebalancing the finger tree as it's traversing it. It could be faster though, I'm not sure.

```
-- anyone could define this and use it today
List.map f xs = 
  go acc = cases
    [] -> acc
    h +: t -> go (acc :+ f h) t
  go []
```

Another option is switching List to some sort of array-mapped trie (branching factor of 32 or 64), so it's effectively constant for indexed access and there's no allocation that needs to happen when walking over it. I kinda like this idea for other reasons and think it's worth trying at some point. Like I think linear traversal of a `Vector` in Scala is about the same speed as traversing a cons list, even though there's technically an extra log base 32 factor.

Any of these changes should probably be guided by profiling rather than just guessing based on asymptotics. Like maybe the list needs to be a gajillion elements before the better asymptotics matter.

One more thing is that when I was looking at `FingerTree`, it does gratuitous allocation even just for lookup. I'm not sure if `Data.Sequence` (which is a specialized `FingerTree`) has the same issue but if so that might be an easy performance win.

**Comment by aryairani** [comment]
Yeah, I forgot about `uncons`/`unsnoc` via pattern matching when I suggested them as builtins.  I dunno what the log base `Data.Sequence` uses, but in the plot you linked (natural log), at 1000 elements it's 6x slower.  Constant factors! \*shakes fist\*
